quic/s2n-quic-core/src/sync/worker.rs

```diff
@@ -0,0 +1,217 @@
+        self.credits -= count;
+    }
+}
+
+/// A handle to submit work to be done to a worker receiver
+///
+/// Multiple Sender handles can be created with `.clone()`.
+#[derive(Clone)]
+pub struct Sender(Arc<State>);
+
+impl Sender {
+    /// Submits `count` jobs to be executed by the worker receiver
+    #[inline]
+    pub fn submit(&self, count: usize) {
+        let state = &*self.0;
+
+        // increment the work counter
+        state.remaining.fetch_add(count, Ordering::Release);
+
+        // wake up the receiver if possible
+        state.receiver.wake();
+    }
+}
+
+impl Drop for Sender {
+    #[inline]
+    fn drop(&mut self) {
+        let state = &*self.0;
+
+        state.senders.fetch_sub(1, Ordering::Release);
+
+        // wake up the receiver to notify that one of the senders has dropped
+        state.receiver.wake();
+    }
+}
+
+struct State {
+    remaining: CachePadded<AtomicUsize>,
+    receiver: AtomicWaker,
+    senders: CachePadded<AtomicUsize>,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            remaining: Default::default(),
+            receiver: Default::default(),
+            senders: AtomicUsize::new(1).into(),
+        }
+    }
+}
+
+struct Acquire<'a>(&'a mut Receiver);
+
+impl<'a> Future for Acquire<'a> {
+    type Output = Option<usize>;
+
+    #[inline]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        self.0.poll_acquire(cx)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::testing::loom;
+
+    fn loom_scenario(iterations: usize, send_batch_size: usize, recv_batch_size: usize) {
+        assert_ne!(send_batch_size, 0);
+        assert_ne!(recv_batch_size, 0);
+
+        loom::model(move || {
+            let (send, mut recv) = channel();
+
+            let sender = loom::thread::spawn(move || {
+                for _ in 0..iterations {
+                    send.submit(send_batch_size);
+                    loom::hint::spin_loop();
+                }
+            });
+
+            let receiver = loom::thread::spawn(move || {
+                loom::future::block_on(async move {
+                    let mut total = 0;
+                    while let Some(mut count) = recv.acquire().await {
+                        assert_ne!(count, 0);
+
+                        while count > 0 {
+                            let to_finish = count.min(recv_batch_size);
+                            recv.finish(to_finish);
+                            total += to_finish;
+                            count -= to_finish;
+                        }
+                    }
+
+                    assert_eq!(total, iterations * send_batch_size);
+                })
+            });
+
+            // loom tests will still run after returning so we don't need to join
+            if cfg!(not(loom)) {
+                sender.join().unwrap();
+                receiver.join().unwrap();
+            }
```

Comment:
what does this mean? how do they keep running after returning?

Reply:
spoke offline: these tests can be run under loom or not. under loom we dont need to join, without it we do.